timemachine/exchange/exchange_mover.py

```diff
@@ -87,108 +95,385 @@ def get_initial_state(water_pdb, mol, ff, seed, nb_cutoff):
+    # (ytz): sample a point inside the sphere uniformly
+    # source: https://karthikkaranth.me/blog/generating-random-points-in-a-sphere/
+    xyz = np.random.randn(3)
+    n = np.linalg.norm(xyz)
+    xyz = xyz / n
+    # (ytz): this might not be inside the box, but it doesnt' matter
+    # since all of our distances are computed using PBCs
+    c = np.cbrt(np.random.rand())
+    new_xyz = xyz * c * radius + center
+    assert np.linalg.norm(delta_r_np(new_xyz, center, box)) < radius
+
+    return new_xyz
+
+
+def v2_insertion(radius, center, box):
+    # generate random proposals outside of the sphere but inside the box
+    while True:
+        xyz = np.random.rand(3) * np.diag(box)
+        if np.linalg.norm(delta_r_np(xyz, center, box)) >= radius:
+            return xyz
+
+
+class TIBDExchangeMove(BDExchangeMove):
+    r"""
+    Targetted Insertion and Biased Deletion Exchange Move
+
+    Insertions are targetted over two regions V1 and V2 such that V1 is a sphere whose origin
+    is at the centroid of a set of indices, and V2 = box - V1.
+
+    Deletions are biased such that each water x_i has a weight equal to exp(u_ixn_nrg(x_i, x \ x_i)).
+    """
+
+    def __init__(
+        self,
+        nb_beta: float,
+        nb_cutoff: float,
+        nb_params: NDArray,
+        water_idxs: NDArray,
+        beta: float,
+        ligand_idxs: List[int],
+        radius: float,
+    ):
+        """
+        Parameters
+        ----
+        nb_beta: float
+            nonbonded beta parameters used in direct space pme
+
+        nb_cutoff: float
+            cutoff in the nonbonded kernel
+
+        nb_params: NDArray (N,4)
+            N is the total number of atoms in the system.
+            Each 4-tuple of nonbonded parameters corresponds to (charge, sigma, epsilon, w coords)
+
+        water_idxs: NDArray (W,3)
+            W is the total number of water molecules in the system.
+            Each element is a 3-tuple denoting the index for oxygen, hydrogen, hydrogen (or whatever is
+            consistent with the nb_params)
+
+        ligand_idxs: List[int]
+            Indices corresponding to the atoms that should be used to compute the centroid
+
+        beta: float
+            Thermodynamic beta, 1/kT (in kJ/mol)
+
+        radius: float
+            Radius to use for the ligand_idxs
+
+        """
+        super().__init__(nb_beta, nb_cutoff, nb_params, water_idxs, beta)
+
+        self.ligand_idxs = np.array(ligand_idxs)  # used to determine center of sphere
+        self.radius = radius
+
+    def get_water_groups(self, coords, box, center):
+        """
+        Partition water molecules into two groups, depending if it's inside the sphere or outside the sphere.
+        """
+        dijs = np.linalg.norm(delta_r_np(np.mean(coords[self.water_idxs_np], axis=1), center, box), axis=1)
+        v1_mols = np.argwhere(dijs < self.radius).reshape(-1)
+        v2_mols = np.argwhere(dijs >= self.radius).reshape(-1)
```

Comment:
nit: readability: move `np.mean` onto its own line `mol_centroids = np.mean(coords[self.water_idxs_np], axis=1)`

(was easy to misread as applying distance test to each atom, rather than each molecule.)

Reply:
willfix